[stable/prometheus-operator] Add support for creating a Service and Ingress for each Prometheus replica

shaunmansell opened the pull request:
<!--
Thank you for contributing to helm/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/helm/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/helm/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the
history. This will make it easier to identify new changes. The PR will be squashed
anyways when it is merged. Thanks.

For fast feedback, please @-mention maintainers that are listed in the Chart.yaml file.

Please make sure you test your changes before you push them. Once pushed, a CircleCI
will run across your changes and do some initial checks and linting. These checks run
very quickly. Please check the results. We would like these checks to pass before we
even continue reviewing your changes.
-->
#### Is this a new chart
> NOTE: We're experiencing a high volume of PRs to this repo and reviews will be delayed. Please host your own chart repository and submit your repository to the Helm Hub instead of this repo to make them discoverable to the community. [Here](https://github.com/helm/hub/blob/master/Repositories.md) is how to submit new chart repositories to the Helm Hub.

No

#### What this PR does / why we need it:
##### What
Adding support for creating separate Service and Ingress resources that will map
to the standard web port of each Prometheus server replica.

We're using the `prometheus.prometheusSpec.replicas` value to determine how many
of each resource type to be created, and to vary resource names and values.

Because we're using a single set of values to control certain properties, we have to
drop support for setting parameters that need to be unique within a each resource.
Primarily this means no IP configuration options for the Service resource.

The ingress resources are created with a single host entry that is compiled by taking
two values (`hostPrefix` and `hostDomain`). The final hostname format is
`hostPrefix`-`replicaIndex`.`hostDomain`, where `replicaIndex` will be 0, 1, 2 etc.

##### Why
My use case for this is to expose each Prometheus server directly via an ingress definition so I can:
- Support Grafana datasources targeting each replica independently. We currently do this via the `prometheus-operated` headless service, however I want to get further metrics on the performance of the inbound requests that Prometheus does not current record internally.
- Scrape each Prometheus server `/metrics` from outside the Kubernetes cluster the server is running in
- Capture requests statistics via the ingress controller metrics to monitor Prometheus API request success and latency

#### Special notes for your reviewer:

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

shaunmansell committed:
Add support for creating a Service and Ingress for each Prometheus replica

Adding support for creating separate Service and Ingress resources that will map
to the standard web port of each Prometheus server replica.

We're using the `prometheus.prometheusSpec.replicas` value to determine how many
of each resource type to be created, and to vary resource names and values.

Because we're using a single set of values to control certain properties, we have to
drop support for setting parameters that need to be unique within a each resource.
Primarily this means no IP configuration options for the Service resource.

The ingress resources are created with a single host entry that is compiled by taking
two values (`hostPrefix` and `hostDomain`). The final hostname format is
`hostPrefix`-`replicaIndex`.`hostDomain`, where `replicaIndex` will be 0, 1, 2 etc.

Signed-off-by: Shaun Mansell <shaun.s.mansell@gmail.com>

shaunmansell committed:
Fix trailing spaces in values and test-values files

Signed-off-by: Shaun Mansell <shaun.s.mansell@gmail.com>

vsliouniaev commented on stable/prometheus-operator/templates/prometheus/serviceperreplica.yaml:
```diff
@@ -0,0 +1,40 @@
+{{- $count := .Values.prometheus.prometheusSpec.replicas | int -}}
+{{- $serviceValues := .Values.prometheus.servicePerReplica -}}
+{{ range $i, $e := until $count }}
+---
```
This should be a list, rather than individual resources

shaunmansell replied:
I've updated both the service and ingress resources to be `ServiceList` and `IngressList` kinds as I took that to be what you meant.

For my own interest, is there any particular reason to prefer these over a templated list of separate resources like what I had originally?

shaunmansell committed:
Merge branch 'master' into prometheus-operator-per-replica-service-and-ingress

shaunmansell committed:
Bump prometheus-operator chart version

Signed-off-by: Shaun Mansell <shaun.s.mansell@gmail.com>

shaunmansell committed:
Convert service and ingress per-replica resources to be managed as a list

Updating both the service and ingress per-replica resources from a templated list to
list api-resource kinds.

Signed-off-by: Shaun Mansell <shaun.s.mansell@gmail.com>
